[Feat] 강의 별 리뷰 조회 기능 구현

sunwook00 opened the pull request:
## ❗️ 이슈 번호
#20 

## 📝 작업 내용
1. Controller 

GET /api/reviews/courses/{courseId} 엔드포인트를 구현.

@PathVariable 및 Pageable을 통해 요청을 파싱.

팀 표준 응답 DTO인 BaseResponse로 응답을 표준화했습니다.


2. Service (ReviewServiceImpl, ReviewService)

Interface와 Impl을 분리하여 설계했습니다.

CourseRepository를 주입받아 courseId로 Course 엔티티를 조회하는 로직을 수행.

Page<Review>(엔티티)를 Page<ReviewResponseDTO>(DTO)로 변환하는 로직을 담당. (.map() 활용)

조회 기능이므로 @Transactional(readOnly = true)로 성능을 최적화했습니다.


3. DTO (ReviewResponseDTO)

Review 엔티티를 기반으로, API 응답에 필요한 데이터(id, content, rating, username, createdAt)만 추출하는 DTO를 생성.


4. Repository (ReviewRepository)

JpaRepository를 상속받았습니다.

Pageable을 파라미터로 받아 동적 정렬 및 페이징을 지원하는 findByCourse 쿼리 메서드를 정의.

## 💭 주의 사항
팀원분들의 조언에 따라, build.gradle의 Spring Security 의존성을 주석 처리하고 테스트를 진행했습니다. (401 오류 회피)

data.sql이 아직 완성되지 않아, Postman 테스트 시 id=1인 강좌가 없어 IllegalArgumentException: 해당 강좌를 찾을 수 없습니다. 
예외가 발생하는 것을 의도대로 확인했습니다.

## 💡 리뷰 포인트
Controller -> Service -> Repository로 이어지는 전체 데이터 흐름과 의존성 주입이 맞는지 확인 부탁드립니다.

Review 엔티티를 ReviewResponseDTO로 변환하는 .map() 로직이 효율적으로 작성되었는지 리뷰 부탁드립니다.

Pageable을 사용한 페이징/정렬 접근 방식이 적절한지 확인 부탁드립니다.

jjeongdong committed:
feat : 수강신청 등록 기능 구현

jjeongdong committed:
feat : 수강 등록 기능 구현

jjeongdong committed:
feat : 수강신청 등록 기능 구현

jjeongdong committed:
Merge conflict resolved

jjeongdong committed:
fix: User Entity 변화로 인한 테스트 코드 수정

sunwook00 committed:
feat: 강의 별 리뷰 조회 기능 추가

sunwook00 committed:
feat: 강의 별 리뷰 조회 기능 추가

sunwook00 committed:
refector: build.gradle 주석 처리 수정

JSeHoone commented on src/main/java/com/example/projectlxp/review/service/ReviewServiceImpl.java:
```diff
@@ -0,0 +1,49 @@
+
+import lombok.RequiredArgsConstructor;
+
+@Service
+@RequiredArgsConstructor // "final이 붙은 필드의 생성자를 자동으로 만들어 줌(생성자 주입)"
+@Transactional(readOnly = true) // 이 클래스의 모든 메서드는 기본적으로 '조회 전용' readOnly = true로 성능 최적화
+public class ReviewServiceImpl implements ReviewService {
```
클래스 단위에서 Transactional을 넣게 되면 해당 서비스의 모든 기능들이 readOnly로 되는걸까요 ?

클래스 내부의 모든 메서드에 적용이 된다면
`createReview`나 `updateReview`가 있는 경우에는 DB에 저장이 안될수도 있을 것 같습니다 !

메서드 단위로 Transactional을 적용해주는 것이 좋아보입니당 !

sunwook00 replied:
네 클래스 위에 Transactional(readOnly = true) 로 어노테이션 해놓으면 해당 클래스의 모든 메서드는 모두 적용되는 걸로 알고있습니다.

sunwook00 replied:
수정 하겠습니다!

jjeongdong replied:
클래스 단위에 `@Transactional(readOnly = true)`을 달아두고
생성 거나 업데이트와 같은 트랜잭션이 필요한 메서드에만 `@Transactional`을 선언하여 사용합니다!

JSeHoone commented on src/main/java/com/example/projectlxp/review/service/ReviewServiceImpl.java:
```diff
@@ -0,0 +1,49 @@
+        // Controller는 courseId만 준다.
+        // Repository에게는 Course(엔티티)를 넘겨줘야 함 그래서 CourseRepository로 먼저 강좌를 찾는다.
+        // 만약 ID에 해당하는 강좌가 없으면, 예외를 발생. (데이터 정합성)
+        Course course =
+                courseRepository
+                        .findById(courseId)
+                        .orElseThrow(
+                                () ->
+                                        new IllegalArgumentException(
+                                                "해당 강좌를 찾을 수 없습니다. id=" + courseId));
```
혹시 이 부분의 들여쓰기가 너무 큰 것 같은데, spotless로 인해서 이렇게 된걸까요 ?..

sunwook00 replied:
아뇨 spotless 사용 안했는데.. 사용하던지 해서 수정하겠습니다 :)

JSeHoone commented on src/main/java/com/example/projectlxp/review/dto/ReviewResponseDTO.java:
```diff
@@ -0,0 +1,21 @@
+
+import com.example.projectlxp.review.entity.Review;
+
+public class ReviewResponseDTO {
+    private Long reviewId;
+    private String content;
+    private int rating;
+    private String username;
+    private LocalDateTime createdAt;
+
+    public ReviewResponseDTO(Review review) {
+        this.reviewId = review.getId();
+        this.content = review.getContent();
+        this.rating = review.getRating();
+        this.username = review.getUser().getName();
+        this.createdAt = review.getCreatedAt();
+    }
+}
```
Getter 메서드가 없어도 API 호출 시 데이터가 잘 출력되나요 ?!..

저는 Getter없으니까 안되서요 !..

sunwook00 replied:
네 저도 방금 집에 와 세훈님이 테스트 데이터 주신 것 넣고 해보니 세훈님이 올려주신 거와 같은 내용으로 406 에러가 떠서 ReviewResponseDTO 클래스 위에 @Getter 어노테이션넣고 global 패키지에 있는 BaseResponse 클래스 위에도 @Getter 어노테이션을 작성 해놨는데 이대로 다시 commit 해도 될까요?!

JSeHoone committed:
feat : Factory Method 추가

- Section 객체를 생성할 때, Factory Method를 이용해서 생성하기

JSeHoone committed:
feat : findByCourseIdAndOrderNo 메서드 추가

- Section내 동일한 CourseId와 OrderNo가 있는지 체크하기 위해서 만들었습니다.

JSeHoone committed:
feat : SectionController - createSection

- Section을 생성하는 API입니다.
- REST URL대로 만들기 위해 노력했습니다.
- Service내 Throw가 발생했을 때, Controller에서 처리합니다.

JSeHoone committed:
feat : Section 생성 시 요청 DTO

- Section 생성 요청이 들어올 때 사용하는 DTO입니다.
- Validation 라이브러리를 사용하여 필수 값이 없으면 에러 응답이 반환되도록 만들었습니다.
- title은 빈 값이 올 수 없도록 설정하였습니다.
- DTO 생성자는 public으로 만들었습니다.

JSeHoone committed:
feat : Section 생성 시 응답 DTO

- Section 생성 후 응답 시 data 필드에 들어가는 값에 대한 DTO입니다.
- Controller에서 최종 응답 시 Data에 직렬화가 되지 않는 이슈를 해결하기 위해 `Getter`어노테이션을 넣어 해결했습니다.
- Validation 라이브러리를 사용하여 필수 값을 받지 않으면 오류가 발생하도록 정의하였습니다.
- title은 빈 값이 올 수 없도록 정의하였습니다.
- 생성자는 public으로 정의하였습니다.

JSeHoone committed:
feat : SectionService interface

- SectionService의 인터페이스를 정의하였습니다.

JSeHoone committed:
feat : SectionServiceImpl

- 인터페이스로 정의한 SectionService를 상속받은 구현체입니다.
- `Service`어노테이션을 사용하여 Bean Context에 등록하였습니다.
- 생성자에 `@Autowired`를 사용하여 의존성을 주입하였습니다.
- registerSection 메서드를 정의하여 섹션이 생성되는 비즈니스 로직을 정의하였습니다.

sunwook00 committed:
refactor: ReviewServiceImpl 클래스 람다 표현식 변환

sunwook00 committed:
refactor: ReviewResponseDTO 클래스 위에 @Getter 어노테이션 추가

sunwook00 committed:
refactor: ReviewServiceImpl 클래스 @Transactional(readOnly = true) 어노테이션 위치 변경

jjeongdong commented on src/main/java/com/example/projectlxp/review/controller/ReviewController.java:
```diff
@@ -0,0 +1,29 @@
+    private final ReviewService reviewService;
+
+    @GetMapping("/courses/{courseId}")
+    public ResponseEntity<BaseResponse<Page<ReviewResponseDTO>>> getReviewsByCourse(
```
ResponseEntity로 안감싸도 괜찮을 것 같습니다!

sunwook00 replied:
넵!

JSeHoone committed:
refactor : ifPresent를 이용하여 throw처리 변경

- findByCourseIdAndOrderNo 메서드에서 ifPresent를 이용하여 IllegalStateException으로 반환되도록 리팩토링

JSeHoone committed:
style : 오타 수정

sunwook00 committed:
FIX: 코드 리뷰 반영(ResponseEntity 래퍼 제거), transactional 어노테이션 위치 변경

sunwook00 committed:
Merge branch 'develop' into feat/#20-강의-별-리뷰-조회-기능-구현
